src/toil/lib/aws/iam.py

```diff
@@ -88,7 +120,49 @@ def check_permission_allowed(perm: str, list_perms: List[str]) -> bool:
             return True
     return False
 
-def allowed_actions_roles(iam: IAMClient, policy_names: List[str], role_name: str) -> Dict[str, List[str]]:
+def get_actions_from_policy_document(policy_doc: Dict[str, Any]) -> AllowedActionCollection:
+    '''
+    Given a policy document, go through each statement and add actions and not actions to an action collection
+
+    :param policy_doc: A policy document to examine
+    '''
+    allowed_actions: AllowedActionCollection = init_action_collection()
+    # Policy document structured like so https://boto3.amazonaws.com/v1/documentation/api/latest/guide/iam-example-policies.html#example
+    logger.debug(policy_doc)
+    for statement in policy_doc["Statement"]:
+
+        if statement["Effect"] == "Allow":
+
+            for resource in statement["Resource"]:
+                for key in ["Action", "NotAction"]:
+                    if key in statement.keys():
+                        if isinstance(statement[key], list):
+                            allowed_actions[resource][key] += statement[key]
+                        else:
+                            #Assumes that if it isn't a list it's probably a string
+                            allowed_actions[resource][key].append(statement[key])
+
+    return allowed_actions
+def allowed_actions_attached(iam: IAMClient, attached_policies: List[AttachedPolicyTypeDef]) -> AllowedActionCollection:
+    """
+    Go through any attached policy we can find to acquire a policy document to get actions from
```

Comment:
Maybe this should say something about how we collect all the grants from all those policies and return them?

It's also written as though we find just one policy document, but actually we look at all the attached policy documents.